src/V3Randomize.cpp

```diff
@@ -196,13 +207,121 @@ class RandomizeMarkVisitor final : public VNVisitor {
             staticRefp->classOrPackagep(VN_AS(staticRefp->varp()->user2p(), NodeModule));
         }
     }
+    // Mark nested MemberSel chain variables as globally constrained
+    void markNestedGlobalConstrainedRecurse(AstNode* nodep) {
+        UASSERT(nodep, "Node should not be null");
+
+        if (VN_IS(nodep, VarRef)) {
+            AstVar* const varp = VN_AS(nodep, VarRef)->varp();
+            if (varp->globalConstrained()) return;
+            varp->globalConstrained(true);
+        } else if (VN_IS(nodep, MemberSel)) {
+            AstMemberSel* const memberSelp = VN_AS(nodep, MemberSel);
+            if (memberSelp->varp()) {
+                AstVar* const varp = memberSelp->varp();
+                if (varp->globalConstrained()) return;
+                varp->globalConstrained(true);
+            }
+            markNestedGlobalConstrainedRecurse(memberSelp->fromp());
+        }
+    }
 
+    // Build MemberSel chain from variable path
+    AstNodeExpr* buildMemberSelChain(AstVarRef* rootVarRefp, const std::vector<AstVar*>& path) {
+        AstNodeExpr* exprp = rootVarRefp->cloneTree(false);
+        for (AstVar* memberVarp : path) {
+            AstMemberSel* memberSelp
+                = new AstMemberSel{rootVarRefp->fileline(), exprp, memberVarp};
+            memberSelp->user2p(m_classp);
+            exprp = memberSelp;
+        }
+        return exprp;
+    }
+
+    // Process a single constraint during nested constraint cloning
+    void processNestedConstraint(AstConstraint* const constrp, AstVarRef* rootVarRefp,
+                                 const std::vector<AstVar*>& newPath) {
+        UASSERT(constrp, "Constraint should not be null");
+
+        AstConstraint* const cloneConstrp = constrp->cloneTree(false);
+
+        std::string pathPrefix = rootVarRefp->name();
+        for (AstVar* pathMemberVarp : newPath) {
+            pathPrefix += GLOBAL_CONSTRAINT_SEPARATOR + pathMemberVarp->name();
+        }
+
+        cloneConstrp->name(pathPrefix + GLOBAL_CONSTRAINT_SEPARATOR + cloneConstrp->name());
+        cloneConstrp->foreach([&](AstVarRef* varRefp) {
+            UASSERT(varRefp && varRefp->varp(), "VarRef should be valid in constraint");
```

Comment:
`foreach` never yields a nullptr, and a VarRef at this stage will always have a `varp`, so this asset is excessive (If those don't hold, there will be mushroom clouds all over the compiler)